lib/Transforms/InsertGpuAllocs.cpp

```diff
@@ -0,0 +1,363 @@
+    mlir::OpBuilder builder(func);
+    llvm::SmallVector<mlir::Value> dims;
+    llvm::SmallPtrSet<mlir::Operation *, 8> filter;
+    auto term = block.getTerminator();
+    assert(term);
+
+    for (auto &it : gpuGetMemrefGlobalParams) {
+      auto getGlobalOp = mlir::cast<mlir::memref::GetGlobalOp>(it.first);
+      it.second = getAccessType(getGlobalOp);
+      it.second.hostRead = true;
+      it.second.hostWrite = true;
+
+      auto loc = getGlobalOp.getLoc();
+      auto access = it.second;
+      builder.setInsertionPointAfter(getGlobalOp);
+      auto getGlobalOpResult = getGlobalOp.getResult();
+      auto memrefType = getGlobalOp.getType().cast<mlir::MemRefType>();
+      auto rank = static_cast<unsigned>(memrefType.getRank());
+      filter.clear();
+      dims.clear();
+      for (auto i : llvm::seq(0u, rank)) {
+        if (memrefType.isDynamicDim(i)) {
+          auto op =
+              builder.create<mlir::memref::DimOp>(loc, getGlobalOpResult, i);
+          dims.push_back(op);
+          filter.insert(op);
+        }
+      }
+      auto allocType = mlir::MemRefType::get(
+          memrefType.getShape(), memrefType.getElementType(),
+          mlir::MemRefLayoutAttrInterface{}, memrefType.getMemorySpace());
+      auto gpuAlloc = builder.create<mlir::gpu::AllocOp>(
+          loc, allocType, /*asyncToken*/ nullptr,
+          /*asyncDependencies*/ llvm::None, dims,
+          /*symbolOperands*/ llvm::None);
+      auto allocResult = gpuAlloc.getResult(0);
+
+      if (access.hostRead || access.hostWrite)
+        gpuAlloc->setAttr(imex::getAllocSharedAttrName(),
+                          builder.getUnitAttr());
+
+      if (access.hostWrite && access.deviceRead) {
+        auto copy =
+            builder.create<mlir::memref::CopyOp>(loc, getGlobalOp, allocResult);
+        filter.insert(copy);
+      }
+
+      if (allocType != memrefType)
+        allocResult =
+            builder.create<mlir::memref::CastOp>(loc, memrefType, allocResult);
+
+      getGlobalOpResult.replaceAllUsesExcept(allocResult, filter);
+
+      builder.setInsertionPoint(term);
+      if (access.hostRead && access.deviceWrite) {
+        builder.create<mlir::memref::CopyOp>(loc, allocResult, getGlobalOp);
+      }
+
+      if (useGpuDealloc)
+        builder.create<mlir::gpu::DeallocOp>(loc, llvm::None, allocResult);
+      else
+        builder.create<mlir::memref::DeallocOp>(loc, allocResult);
+    }
+
+    // This is the case where a memref.alloc op is directly converted to
+    // gpu.alloc
+    for (auto it : gpuBufferAllocs) {
+      auto alloc = mlir::cast<mlir::memref::AllocOp>(it.first);
+      auto access = it.second;
+      auto loc = alloc.getLoc();
+      builder.setInsertionPoint(alloc);
+      auto gpuAlloc = builder.create<mlir::gpu::AllocOp>(
+          loc, alloc.getType(), /*asyncToken*/ nullptr,
+          /*asyncDependencies*/ llvm::None, alloc.dynamicSizes(),
```

Comment:
Why the getType and size are different than the global and parameters?  

      for (auto i : llvm::seq(0u, rank)) {
        if (memrefType.isDynamicDim(i)) {
          auto op = builder.create<mlir::memref::DimOp>(loc, param, i);
          dims.push_back(op);
          filter.insert(op);
        }
      }
      auto allocType = mlir::MemRefType::get(
          memrefType.getShape(), memrefType.getElementType(),
          mlir::MemRefLayoutAttrInterface{}, memrefType.getMemorySpace());
      auto gpuAlloc = builder.create<mlir::gpu::AllocOp>(
          loc, allocType, /*asyncToken*/ nullptr,
          /*asyncDependencies*/ llvm::None, dims,
          /*symbolOperands*/ llvm::None);

Reply:
because here you directly have a memref alloc to convert to gpu alloc and global and params case you dont have it, you add the allocs and subsequent mem copy's